diffusion/models/precomputed_text_latent_diffusion.py

```diff
@@ -0,0 +1,496 @@
+        clip_attn_mask = clip_tokenizer_out['attention_mask'].to(torch.bool).to(device)
+        clip_out = self.clip_encoder(input_ids=clip_tokenized_captions,
+                                     attention_mask=clip_attn_mask,
+                                     output_hidden_states=True)
+        clip_embed = clip_out.hidden_states[-2]
+        pooled_embeddings = clip_out[1]
+        return t5_embed, clip_embed, t5_attn_mask, clip_attn_mask, pooled_embeddings
+
+    def prepare_text_embeddings(self, t5_embed: torch.Tensor, clip_embed: torch.Tensor, t5_mask: torch.Tensor,
+                                clip_mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        if t5_embed.shape[1] > self.max_seq_len:
+            t5_embed = t5_embed[:, :self.max_seq_len]
+            t5_mask = t5_mask[:, :self.max_seq_len]
+        if clip_embed.shape[1] > self.max_seq_len:
+            clip_embed = clip_embed[:, :self.max_seq_len]
+            clip_mask = clip_mask[:, :self.max_seq_len]
+        t5_embed = self.t5_proj(t5_embed)
+        clip_embed = self.clip_proj(clip_embed)
+        # Add position embeddings
+        t5_embed = 0.707 * t5_embed + 0.707 * self.t5_position_embedding[:t5_embed.shape[1]].unsqueeze(0)
+        clip_embed = 0.707 * clip_embed + 0.707 * self.clip_position_embedding[:clip_embed.shape[1]].unsqueeze(0)
+        # Apply layernorms
+        t5_embed = self.t5_ln(t5_embed)
+        clip_embed = self.clip_ln(clip_embed)
+        # Concatenate the text embeddings
+        text_embeds = torch.cat([t5_embed, clip_embed], dim=1)
+        encoder_attention_mask = torch.cat([t5_mask, clip_mask], dim=1)
+        return text_embeds, encoder_attention_mask
+
+    def diffusion_forward(self, latents, timesteps):
+        # Add noise to the inputs (forward diffusion)
+        noise = torch.randn(*latents.shape, device=latents.device, generator=self.rng_generator)
+        noised_latents = self.noise_scheduler.add_noise(latents, noise, timesteps)
+        # Generate the targets
+        if self.prediction_type == 'epsilon':
+            targets = noise
+        elif self.prediction_type == 'sample':
+            targets = latents
+        elif self.prediction_type == 'v_prediction':
+            targets = self.noise_scheduler.get_velocity(latents, noise, timesteps)
+        else:
+            raise ValueError(
+                f'prediction type must be one of sample, epsilon, or v_prediction. Got {self.prediction_type}')
+        return noised_latents, targets
+
+    def forward(self, batch):
+        latents, text_embeds, text_pooled_embeds, encoder_attention_mask = None, None, None, None
+
+        # Encode the images with the autoencoder encoder
+        inputs = batch['image']
+        latents = self.encode_images(inputs)
+
+        # Text embeddings are shape (B, seq_len, emb_dim), optionally truncate to a max length
+        t5_embed = batch['T5_LATENTS']
```

Comment:
just to write comments consistent with what i said for the dataloader, probably safest to grab these key names from an argument but this is totally fine for our use cases

Reply:
Changed it to get key names from an argument.